File: Leetcode/20-Valid-Parentheses/61-Rotate-List/61-Rotate-List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == nullptr || head->next == nullptr) {
            return head;
        }
        
        int length = 1;
        ListNode *temp = head;

        while (temp->next != nullptr) {
            temp = temp->next;
            length++;
        }

        temp->next = head;
        temp = head;

        int counter = length - (k % length);
        for (int i = 1; i < counter; i++) {
            temp = temp->next;
        }

        head = temp->next;
        temp->next = nullptr;
        return head;
    }
};

File: Leetcode/61-Rotate-List/61-Rotate-List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode temp = head, tail = head;
        int length = 0;

        while (temp != null) {
            if (temp.next == null) {
                tail = temp;
            }

            temp = temp.next;
            length++;
        }

        k = k % length;

        System.out.println(length + " " + k);

        if (k == 0) {
            return head;
        }

        ListNode newTail = head;

        for (int i = 0; i < length - k - 1; i++) {
            newTail = newTail.next;
        }

        ListNode newHead = newTail.next;
        newTail.next = null;
        tail.next = head;

        return newHead;
    }
}